File: internal/bills/allocation.go
package bills

import "sort"

func allocate(total Cents, shares []Share) []AllocatedShare {
	type remainder struct {
		position int
		value    int64
	}

	allocated := make([]AllocatedShare, len(shares))
	remainders := make([]remainder, len(shares))
	var allocatedTotal Cents
	for index, share := range shares {
		productWhole := int64(total/10_000) * int64(share.Percentage)
		productRemainder := int64(total%10_000) * int64(share.Percentage)
		amount := Cents(productWhole + productRemainder/10_000)
		allocated[index] = AllocatedShare{
			PersonName: share.PersonName,
			Percentage: share.Percentage,
			Amount:     amount,
		}
		allocatedTotal += amount
		remainders[index] = remainder{position: index, value: productRemainder % 10_000}
	}

	sort.Slice(remainders, func(i, j int) bool {
		if remainders[i].value == remainders[j].value {
			return remainders[i].position < remainders[j].position
		}
		return remainders[i].value > remainders[j].value
	})
	for index := Cents(0); index < total-allocatedTotal; index++ {
		allocated[remainders[index].position].Amount++
	}

	return allocated
}

File: frontend/src/allocation.ts
import { formatAmount, parseAmount, parsePercentage } from "./decimal"
import type { DraftShare } from "./types"

interface Remainder {
  position: number
  value: bigint
}

export function allocateDraft(
  totalAmount: string,
  shares: DraftShare[],
): Array<string | null> {
  const totalCents = parseAmount(totalAmount)
  if (totalCents === null) {
    return shares.map(() => null)
  }

  const percentages = shares.map((share) => parsePercentage(share.percentage))
  const amounts: Array<bigint | null> = []
  const remainders: Remainder[] = []
  let allocatedTotal = 0n

  percentages.forEach((percentage, position) => {
    if (percentage === null) {
      amounts.push(null)
      return
    }

    const product = totalCents * BigInt(percentage)
    const amount = product / 10_000n
    amounts.push(amount)
    allocatedTotal += amount
    remainders.push({ position, value: product % 10_000n })
  })

  const percentageTotal = percentages.reduce<number>(
    (total, percentage) => total + (percentage ?? 0),
    0,
  )
  if (percentageTotal === 10_000 && remainders.length === shares.length) {
    remainders.sort((left, right) => {
      if (left.value === right.value) {
        return left.position - right.position
      }
      return left.value > right.value ? -1 : 1
    })
    const remainingCents = Number(totalCents - allocatedTotal)
    for (let index = 0; index < remainingCents; index += 1) {
      const position = remainders[index].position
      amounts[position] = (amounts[position] ?? 0n) + 1n
    }
  }

  return amounts.map((amount) => (amount === null ? null : formatAmount(amount)))
}
